src/stan/mcmc/chainset.hpp

```diff
@@ -0,0 +1,504 @@
+
+  /**
+   * Compute the quantile value of the specified parameter
+   * at the specified probability.
+   *
+   * Throws exception if specified probability is not between 0 and 1.
+   *
+   * @param index parameter index
+   * @param prob probability
+   * @return parameter value at quantile
+   */
+  double quantile(const int index, const double prob) const {
+    // Ensure the probability is within [0, 1]
+    if (prob <= 0.0 || prob >= 1.0) {
+      throw std::out_of_range("Probability must be between 0 and 1.");
+    }
+    Eigen::MatrixXd draws = samples(index);
+    Eigen::Map<Eigen::VectorXd> map(draws.data(), draws.size());
+    return stan::math::quantile(map, prob);
+  }
+
+  /**
+   * Compute the quantile value of the specified parameter
+   * at the specified probability.
+   *
+   * Throws exception if specified probability is not between 0 and 1.
+   *
+   * @param name parameter name
+   * @param prob probability
+   * @return parameter value at quantile
+   */
+  double quantile(const std::string& name, const double prob) const {
+    return quantile(index(name), prob);
+  }
+
+  /**
+   * Compute the quantile values of the specified parameter
+   * for a set of specified probabilities.
+   *
+   * Throws exception if any probability is not between 0 and 1.
+   *
+   * @param index parameter index
+   * @param probs vector of probabilities
+   * @return vector of parameter values for quantiles
+   */
+  Eigen::VectorXd quantiles(const int index,
+                            const Eigen::VectorXd& probs) const {
+    if (probs.size() == 0)
+      return Eigen::VectorXd::Zero(0);
+    if (probs.minCoeff() <= 0.0 || probs.maxCoeff() >= 1.0) {
+      throw std::out_of_range("Probabilities must be between 0 and 1.");
+    }
+    Eigen::MatrixXd draws = samples(index);
+    Eigen::Map<Eigen::VectorXd> map(draws.data(), draws.size());
+    std::vector<double> probs_vec(probs.data(), probs.data() + probs.size());
+    std::vector<double> quantiles = stan::math::quantile(map, probs_vec);
+    return Eigen::Map<Eigen::VectorXd>(quantiles.data(), quantiles.size());
+  }
+
+  /**
+   * Compute the quantile values of the specified parameter
+   * for a set of specified probabilities.
+   *
+   * Throws exception if any probability is not between 0 and 1.
+   *
+   * @param name parameter name
+   * @param probs vector of probabilities
+   * @return vector of parameter values for quantiles
+   */
+  Eigen::VectorXd quantiles(const std::string& name,
+                            const Eigen::VectorXd& probs) const {
+    return quantiles(index(name), probs);
+  }
+
+  /**
+   * Computes the split potential scale reduction (split Rhat) using rank based
+   * diagnostic for a set of per-chain draws, for bulk and tail Rhat.
+   * Based on paper https://arxiv.org/abs/1903.08008
+   *
+   * @param index parameter index
+   * @return pair (bulk_rhat, tail_rhat)
+   */
+  std::pair<double, double> split_rank_normalized_rhat(const int index) const {
```

Comment:
(I still think it is worth adding functions for the "old" rhat and ess to this class, if they'd be this same size)